RFC: Interrupt calling conventions

phil-opp opened the pull request:
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rfcs/pull/3246)*

Add compiler support for interrupt calling conventions that are specific to an architecture or target. This way, interrupt handler functions can be written directly in Rust without needing assembly shims.

While there are already some unstable implementations of interrupt calling conventions in `rustc`, they were [only added for experimentation](https://github.com/rust-lang/rust/pull/38465#issuecomment-270015470). As these calling conventions showed their worth over the years, the lang team [requested an RFC](https://github.com/rust-lang/rust/issues/40180#issuecomment-1022507941) for interrupt calling conventions general, to put them on a path towards stabilization. So this RFC tries to provide the necessary groundwork for such calling conventions.

[Rendered](https://github.com/phil-opp/rfcs/blob/interrupt-calling-conventions/text/0000-interrupt-calling-conventions.md)

[Pre-RFC discussion on internals](https://internals.rust-lang.org/t/pre-rfc-interrupt-calling-conventions/16182)

phil-opp committed:
Start the interrupt calling convention RFC

phil-opp committed:
First draft on motivation and guide-level explanation sections

phil-opp committed:
Write reference-level explanation, drawbacks, and alternatives section

phil-opp committed:
Prior art, open questions, and future possibilities

phil-opp committed:
Fix typos

phil-opp committed:
Various improvements after proof-reading

phil-opp committed:
Stabilizations for tier 3 should only be done with caution

phil-opp committed:
Clarify the stability and platform support of interrupt calling conventions

phil-opp committed:
Some architectures might have multiple different interrupt calling conventions

nikomatsakis commented:
To be sure I understand, this RFC is not proposing any specific interrupt ABI, but simply talking about the *idea* of interrupt handler ABIs and what their requirements would be?

bjorn3 commented on text/0000-interrupt-calling-conventions.md:
````diff
@@ -0,0 +1,181 @@
+
+  ```rust
+  extern "x86-interrupt" fn(stack_frame: &ExceptionStackFrame);
+  extern "x86-interrupt" fn(stack_frame: &ExceptionStackFrame, error_code: u64);
````
Is the error code also 64bit on 32bit systems?

phil-opp replied:
Good catch! I pushed a [new commit](https://github.com/rust-lang/rfcs/pull/3246/commits/719b02c30361106d0523a2b15c816a618103fd94) with some more details on this.

bjorn3 commented on text/0000-interrupt-calling-conventions.md:
```diff
@@ -0,0 +1,181 @@
+## Stability
+Interrupt calling conventions are a language feature, so they fall under Rust's normal stability guarantees, with one exception: If official support for a target is dropped, the corresponding interrupt calling convention can be removed from the Rust language, even if it is stabilized. This is not considered a breaking change because no code on other targets is broken by this, since it was never p ...
+
+As soon as a calling convention is stabilized, breaking changes are longer allowed, independent of the target tier. For this reason, special caution must be taken before stabilizing interrupt calling conventions for _tier 3_ targets, as these targets might still evolve. Special care must also be taken before stabilizing interrupt calling conventions that are implemented outside of `rustc` (e.g. i ...
```
*are _no_ longer

phil-opp commented:
> To be sure I understand, this RFC is not proposing any specific interrupt ABI, but simply talking about the _idea_ of interrupt handler ABIs and what their requirements would be?

Yes, exactly. I added some more context to the PR description. The plan is to use this RFC as a foundation, so that we don't need separate RFCs for each individual interrupt ABI. See @joshtriplett's comment in https://github.com/rust-lang/rust/issues/40180#issuecomment-1022507941 for more details.

phil-opp committed:
Fix typo

phil-opp committed:
Mention `rust_codegen_gcc`

phil-opp committed:
Add more details on x86-interrupt calling convention

- the error code is only u64 on 64-bit systems, on 32-bit systems it needs to be u32
- mention the Rust compiler doesn't check these types yet

phil-opp commented on text/0000-interrupt-calling-conventions.md:
```diff
@@ -0,0 +1,181 @@
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Interrupt calling conventions can be quite complex. So even though they are a very isolated feature, they still **add a considerable amount of complexity to the Rust language**. This added complexity could lead to considerable work for alternative Rust compilers/code generators that don't build on top of LLVM. Examples are [`cranelift`](https://github.com/bytecodealliance/wasmtime/tree/main/crane ...
```
Thanks, added!

bjorn3 commented on text/0000-interrupt-calling-conventions.md:
````diff
@@ -78,10 +78,12 @@ The following interrupt ABIs are currently supported:
   ```
   The `error_code` argument is _not_ an optional argument. It is set by the hardware for some interrupt vector, but not for others. The programmer must make sure to always use the correct signature for each interrupt vector, otherwise undefined behavior occurs.
+
+  The `StackFrame` type must be a struct that matches the stack frame pushed by the CPU. The `ErrorCode` type must be `u64` on 64-bit targets and `u32` on 32-bit targets. These types are currently _not_ checked by `rustc`.
````
Maybe use usize as error code type instead?

tbu- replied:
I think it makes sense being explicit here. `usize` isn't necessarily defined to be the width of a register.

nikomatsakis commented:
I'm in favor of this. I guess this is a @rust-lang/lang decision, but I would want @rust-lang/compiler input too!

fweimer commented on text/0000-interrupt-calling-conventions.md:
```diff
@@ -78,10 +78,12 @@ The following interrupt ABIs are currently supported:
-  extern "x86-interrupt" fn(stack_frame: &ExceptionStackFrame);
-  extern "x86-interrupt" fn(stack_frame: &ExceptionStackFrame, error_code: u64);
+  extern "x86-interrupt" fn(stack_frame: &StackFrame);
+  extern "x86-interrupt" fn(stack_frame: &StackFrame, error_code: ErrorCode);
```
This needs to specify what exactly is being saved on the stack, and how the interrupt is exited.

x86 supports various different asynchronous interrupts, and based on the description, it's not clear which one is supported by these calling conventions.

phil-opp replied:
These calling conventions are listed only as an example. So I think it makes more sense to specify such details elsewhere, e.g. in the future stabilization proposals for the respective calling convention.

fweimer commented on text/0000-interrupt-calling-conventions.md:
```diff
@@ -89,61 +89,67 @@ By using these ABIs, it is possible to implement interrupt handlers directly in
+
+Apart from this limitation, interrupt calling conventions fall under Rust's normal stability guarantees. For this reason, special care must be taken before stabilizing interrupt calling conventions that are implemented outside of `rustc` (e.g. in LLVM).
+
+## Safety
```
This needs to say that interrupt routines are implicitly unsafe. Even if the interrupt routine only calls safe Rust functions, it can easily observe intermediate CPU states that are beyond the Rust safety model, or corrupt the state so that undefined behavior occurs after the interrupt routine returns.

phil-opp replied:
Could you clarify what you mean by "implictly unsafe"? That they should be treated like an `unsafe fn` so that no `unsafe {}` blocks are required in it?

I agree this question is something that we should specify in the RFC in more detail. My opinion on this is that we should keep the function body safe, but make all dangerous operations unsafe:

- Functions with interrupt calling conventions cannot be called from Rust code directly, so there is no way to cause undefined behavior using only safe code.
- To register a interrupt handler function on the CPU, you typically need some unsafe code (e.g. inline asm) or place it at a specific address using a linker script (which allow all kinds of unsafe operations). In this "unsafe" code, you need to make sure that the referenced function fulfills all necessary safety invariants.
- We should try to make all operations that corrupt the CPU state unsafe, if possible. For example, the `x86_64` crate requires an `unsafe` block to [modify the passed `InterruptStackFrame`](https://docs.rs/x86_64/0.14.8/x86_64/structures/idt/struct.InterruptStackFrame.html#method.as_mut). Ideally, this `unsafe` block should be required directly by the compiler, for example by requiring raw pointer arguments instead of references for the `x86-interrupt` calling convention.

I'm happy to discuss this further!

fweimer replied:
What I meant is that even if the function is not marked as unsafe, the usual rules about Rust memory safety do not apply at all to these functions. There are a lot of restrictions that have to be followed, depending on the rest of the system, and I'm not sure if Rust is actually up for that. (With C or C++, separate compilation with different compiler flags is sometimes used.)

Lokathor replied:
That seems like a case by case basis as to how safe each ABI is going to be. The interrupt handler function ABI *could* (for example) have additional pre and post code inserted by rustc automatically so that the user's source code always runs in a safe environment.

Like we can literally make up these ABIs to do whatever, since normal code is never going to interface with the interrupt handler code.

nikomatsakis replied:
@fweimer can you elaborate a bit more about what rules re: Rust memory safety are inapplicable? The function doesn't get access to anything but static data, right?

Lokathor replied:
The function in fact *could* get access to the register values the main program had when the interrupt happened. The function also *might* get far far less stack space than normal, and would likely not have guard pages set up for this stack. There's a lot of potential interrupt styles.

nikomatsakis replied:
@Lokathor how could it get access to those things, specifically?

Lokathor replied:
This is quite "into the weeds", but for example on ARM 32-bit there's multiple CPU modes, and an interrupt will make the CPU switch modes and jump to the interrupt handler, but separate modes *mostly* share all the same registers. So your interrupt handler just naturally sees the same registers as the main program happened to have at the time the interrupt happened. In rust terms this doesn't involve unsafe at all. In fact, for a software interrupt this is intentional. However, this does mean that the handler has to properly preserve all register values and restore them before ending the interrupt.

Lokathor replied:
Ah, I see your point. Well then at least the situations I was talking about are fine enough if the compiler is handling it. Though, perhaps **fweimer** had something else in mind.

bjorn3 replied:
Once you introduce interrupts you are effectively no longer a single-threaded process as the easiest way to model interrupts in the rust abstract machine is as running on a separate thread which could synchronize with the non-interrupt thread using regular concurrency methods. It just so happens that while the interrupt runs, the non-interrupt thread doesn't make any progress and thus you may deadlock if you wait for it.

chorman0773 replied:
I don't think it can be treated the same as a separate thread precisely because it blocks the non-interrupt code while running.

I've equated it to a signal handler before, which is far more limited in what it can do.

mark-i-m commented:
I still think I would rather have custom calling conventions in general (i.e., allow me to define my own calling convention and use it). There are a few reasons I think a more general mechanism might be better:
* I think that could play nicely with naked functions.
* It would also allow defining architecture-specific conventions in libraries (e.g., as macros) rather than hardcoding them all into the compiler.
* It would allow people to use custom conventions for things that are not interrupts (e.g., trap and syscall handlers, userspace thread switchers, signal handlers, etc).

bjorn3 commented:
While that is a nice idea IMO, neither LLVM, nor GCC, nor Cranelift support that at the moment.

phil-opp commented:
> It seems buildable with naked functions and inline assembly.

You only need to save the registers that are overwritten by called functions. The register selection happens in LLVM, so it's not possible to know this earlier. So using naked functions, you need to save and restore all registers, which is more expensive.

mark-i-m commented:
@phil-opp I guess my expectation is that anything using a custom calling convention would not be callable from Rust. Moreover, the calling convention would specify who needs to save and restore what registers anyway.

phil-opp commented:
> Moreover, the calling convention would specify who needs to save and restore what registers anyway.

But how can we implement this without LLVM support?

mark-i-m commented:
I imagine something like this (bikesheddable syntax):

```rust
// I haven't fully though through what this would look like...
#[calling_convention(
  name="mark-weird-exception-handler",
  arch="x86_64",
  // custom prologue and epilogue allow setting up the stack or any other state however you want...
  prologue = asm!(...),
  epilogue = asm!(...),
  args=["regs: Regs", "error_code: u64"], // function with this convention must have these arguments
)]

#[no_mangle]
fn extern "mark-weird-exception-handler" handler(regs: Regs, error_code: u64) {
  println!("{regs} {error_code:X}");
  // other code
}
```

would desugar to something like this:

```rust
#[naked]
#[no_mangle]
fn handler() {
  <prologue>

  asm!("call _handler_inner");

  <epilogue>
}

#[no_mangle]
fn extern "C" _handler_inner(regs: Regs, error_code: u64) {
   println!("{regs} {error_code:X}");
  // other code
}
```

The basic idea is that the custom calling convention is tasked with converting a weird calling convention into something the compiler can work with normally (e.g., the C or Rust calling conventions, though I prefer C, since it is specified and gives more control). I think this mostly means fixing up the stack, but it could also include other stateful things, such as enabling/disabling interrupts, fixing up page tables, etc.

The custom calling convention then becomes a simple trampoline mechanism into more normal Rust code.

phil-opp commented:
The issue with this approach is that you have to write the assembly for the prologue and epilogue yourself. This includes saving and restoring all registers (on many platforms). To do that in an optimal way, you need to know which registers are actually overwritten by `_handler_inner`. However, this is not decided until LLVM performs [register allocation](https://en.wikipedia.org/wiki/Register_allocation). So without LLVM support you need to save _all_ registers instead.

One way to work around that problem could be to use LLVM's `preserve_all` calling convention for your `_handler_inner` function. This way, LLVM would generate code that automatically restores all registers to their original values before returning,  so we would not need to do it manually anymore. The RFC already describes this idea in the "rationale and alternatives" section at the end. I'm happy to explore and discuss this approach further!

phil-opp commented:
I don't have any realistic use cases, but you can easily construct an example where the LLVM-generated code is much better:

```rust
static TICKS: AtomicUsize = AtomicUsize::new(0);
const THRESHOLD: usize = 10000;

#[no_mangle]
pub extern "x86-interrupt" fn timer_handler(_stack_frame: u8) {
    let prev = TICKS.fetch_add(1, Ordering::Acquire);
    if prev > THRESHOLD {
        // trigger a "context switch" interrupt 
        unsafe { asm!("int 0xf0"); }
    }
}
```

We want to count the timer ticks with minimal overhead. If a specified threshold value is passed, we trigger a secondary interrupt, which will save the registers and perform a context switch.

When you try this example [on godbolt](https://rust.godbolt.org/z/YTvKPTv69), you see that `rustc` generates the following assembly:

```asm
timer_handler:
        push    rax
        cld
        mov     eax, 1
        lock            xadd    qword ptr [rip + example::TICKS], rax
        cmp     rax, 10000
        jbe     .LBB0_1

        int     240

        pop     rax
        iretq
.LBB0_1:
        pop     rax
        iretq

example::TICKS:
        .zero   8
```

So we only save and restore the `rax` register, which leads to a much lower overhead, especially if the timer interrupt fires frequently.

Lokathor commented:
We certainly should consider "some sort of embedded use case" to be the *primary* use case with this sort of thing. In which case all register savings are a big win.

phil-opp commented:
> We certainly should consider "some sort of embedded use case" to be the primary use case with this sort of thing. In which case all register savings are a big win.

Yes, I agree! Does someone have a similar example for some embedded device? Then I'd be happy to extend the motivation section of the RFC based on that.

phil-opp committed:
Extend paragraph about register backup/restore in motivation

Lokathor commented:
On the GBA we're using an assembly interrupt handler which reads a nullable function pointer to an `extern "C" fn(u16)` (which is presumably written in Rust). If the assembly reads a non-null value then it has to mode switch, align the stack in system mode, save some registers, do the call, and then undo all that. If a custom handler ABI was allowed for use we could just skip over the register saving and stack alignment steps and have LLVM handle any necessary work in each particular `extern "gba-irq" fn(u16)` implementation.

chorman0773 commented:
I posted this on the tracking issue but it makes sense to cross-post to the RFC.

In 32-bit mode, the StackFrame struct can vary dynamically - depending on the call cs.CPL (which may or may not end up in saveCS[1:0]). Specifically, ss:eSP are only pushed by the CPU when calling a lower privileged gate, and only popped when returning to higher privileged code (gate.selector.RPL<CPL, and CPL<cs.RPL respectively - note that it's impossible jump to a higher privileged gate, nor iret to lower privileged code).

Nyabinary commented:
What's the status of this RFC?

natevw commented:
Should the RISC-V interrupt (<https://github.com/rust-lang/rust/issues/111889>) be added to the final paragraph of the Motivations section as well?

Lancelotbronner commented:
What's the status on aarch64 interrupt abi?
